In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from sklearn.preprocessing import StandardScaler

In [2]:
df=pd.read_csv('processed.cleveland.data',header=None)


      1. age      
      2. sex     
      3. cp - chest pain      
      4. restbps - resting blood pressure (in mm Hg) 
      5. chol - cholesterol     
      6. fbs - fasting blood sugar       
      7. restecg -resting electrocardiographic results   
      8. thalach - maximum heart rate achieved   
      9. exang - exercise induced angina     
      10. oldpeak - ST depression induced by exercise relative to rest   
      11. slope - the slope of the peak exercise ST segment    
      12. ca -  number of major vessels (0-3) colored by flourosopy       
      13. thal - thallium heart scan  
      14. hd - diagnosis of heart disease (the predicted attribute)

In [3]:
df.columns=['age',
            'sex',
            'cp',
            'restbps',
            'chol',
            'fbs',
            'restecg',
            'thalach',
            'exang',
            'oldpeak',
            'slope',
            'ca',
            'thal',
            'hd']

In [4]:
len(df.loc[(df['ca']=='?') | (df['thal']=='?')])

6

In [5]:
df.loc[(df['ca']=='?') | (df['thal']=='?')]

age  sex   cp  restbps   chol  fbs  restecg  thalach  exang  oldpeak  \
87   53.0  0.0  3.0    128.0  216.0  0.0      2.0    115.0    0.0      0.0   
166  52.0  1.0  3.0    138.0  223.0  0.0      0.0    169.0    0.0      0.0   
192  43.0  1.0  4.0    132.0  247.0  1.0      2.0    143.0    1.0      0.1   
266  52.0  1.0  4.0    128.0  204.0  1.0      0.0    156.0    1.0      1.0   
287  58.0  1.0  2.0    125.0  220.0  0.0      0.0    144.0    0.0      0.4   
302  38.0  1.0  3.0    138.0  175.0  0.0      0.0    173.0    0.0      0.0   

     slope   ca thal  hd  
87     1.0  0.0    ?   0  
166    1.0    ?  3.0   0  
192    2.0    ?  7.0   1  
266    2.0  0.0    ?   2  
287    2.0    ?  7.0   0  
302    1.0    ?  3.0   0

In [6]:
df_new=df.loc[(df['ca']!='?') & (df['thal']!='?')]

In [7]:
X=df_new.drop('hd',axis=1).copy()
X.head()

age  sex   cp  restbps   chol  fbs  restecg  thalach  exang  oldpeak  \
0  63.0  1.0  1.0    145.0  233.0  1.0      2.0    150.0    0.0      2.3   
1  67.0  1.0  4.0    160.0  286.0  0.0      2.0    108.0    1.0      1.5   
2  67.0  1.0  4.0    120.0  229.0  0.0      2.0    129.0    1.0      2.6   
3  37.0  1.0  3.0    130.0  250.0  0.0      0.0    187.0    0.0      3.5   
4  41.0  0.0  2.0    130.0  204.0  0.0      2.0    172.0    0.0      1.4   

   slope   ca thal  
0    3.0  0.0  6.0  
1    2.0  3.0  3.0  
2    2.0  2.0  7.0  
3    3.0  0.0  3.0  
4    1.0  0.0  3.0

In [176]:
y=df_new['hd'].copy()

In [177]:
y_not_zero=y>0
#get the index for each non zero value in y
y[y_not_zero]=1 
#set them to 1
y.unique()

array([0, 1], dtype=int64)

age: age in years, FLOAT

sex: sex (1 = male; 0 = female), CATEGORICAL

cp: chest pain type, CATEGORICAL

   -- Value 1: typical angina
   -- Value 2: atypical angina
   -- Value 3: non-anginal pain
   -- Value 4: asymptomatic 

restbps: resting blood pressure (in mm Hg), FLOAT

chol: serum cholestoral in mg/dl, FLOAT

fbs: (fasting blood sugar > 120 mg/dl)  (1 = true; 0 = false), CATEGORICAL

restecg: resting electrocardiographic results, CATEGORICAL

   -- Value 0: normal
   -- Value 1: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)
   -- Value 2: showing probable or definite left ventricular hypertrophy by Estes' criteria

thalach: maximum heart rate achieved, FLOAT

exang: exercise induced angina (1 = yes; 0 = no), CATEGORICAL

oldpeak = ST depression induced by exercise relative to rest, FLOAT

slope: the slope of the peak exercise ST segment, CATEGORICAL

   -- Value 1: upsloping
   -- Value 2: flat
   -- Value 3: downsloping

ca: number of major vessels (0-3) colored by flourosopy, FLOAT

thal: thallium heart scan, CATEGORICAL

   -- 3 = normal
   -- 6 = fixed defect
   -- 7 = reversable defect


In [183]:
X_encoded=pd.get_dummies(X,columns=['cp','restecg','slope','thal'],dtype='float64')
#default dtype is bool
X_encoded.head()

age  sex  restbps   chol  fbs  thalach  exang  oldpeak   ca  cp_1.0  ...  \
0  63.0  1.0    145.0  233.0  1.0    150.0    0.0      2.3  0.0     1.0  ...   
1  67.0  1.0    160.0  286.0  0.0    108.0    1.0      1.5  3.0     0.0  ...   
2  67.0  1.0    120.0  229.0  0.0    129.0    1.0      2.6  2.0     0.0  ...   
3  37.0  1.0    130.0  250.0  0.0    187.0    0.0      3.5  0.0     0.0  ...   
4  41.0  0.0    130.0  204.0  0.0    172.0    0.0      1.4  0.0     0.0  ...   

   cp_4.0  restecg_0.0  restecg_1.0  restecg_2.0  slope_1.0  slope_2.0  \
0     0.0          0.0          0.0          1.0        0.0        0.0   
1     1.0          0.0          0.0          1.0        0.0        1.0   
2     1.0          0.0          0.0          1.0        0.0        1.0   
3     0.0          1.0          0.0          0.0        0.0        0.0   
4     0.0          0.0          0.0          1.0        1.0        0.0   

   slope_3.0  thal_3.0  thal_6.0  thal_7.0  
0        1.0       0.0       1.0       0.0  
1        0.0       1.0       0.0       0.0  
2        0.0       0.0       0.0       1.0  
3        1.0       1.0       0.0       0.0  
4        0.0       1.0       0.0       0.0  

[5 rows x 22 columns]

In [184]:
X_train,X_test,y_train,y_test=train_test_split(X_encoded,y,random_state=42)


In [ ]:

scaler = StandardScaler()
cols_to_scale = ['age', 'restbps', 'chol', 'thalach', 'oldpeak', 'ca']  

X_train[cols_to_scale] = scaler.fit_transform(X_train[cols_to_scale])
X_test[cols_to_scale] = scaler.transform(X_test[cols_to_scale])


In [186]:
X_train[:5]

age  sex   restbps      chol  fbs   thalach  exang   oldpeak  \
66   0.624570  1.0  0.464737 -1.191504  0.0  0.282152    0.0  1.746939   
105 -0.044720  1.0 -1.345292  1.134058  0.0  0.324647    0.0 -0.914226   
160 -0.937106  1.0 -1.741235 -0.966450  1.0  0.324647    0.0 -0.914226   
147 -1.494848  1.0 -1.119038  0.027540  0.0  1.302033    0.0 -0.914226   
19  -0.602461  1.0 -0.100897  0.327613  0.0  0.962073    0.0 -0.381993   

           ca  cp_1.0  ...  cp_4.0  restecg_0.0  restecg_1.0  restecg_2.0  \
66  -0.709045     0.0  ...     0.0          0.0          0.0          1.0   
105 -0.709045     0.0  ...     0.0          1.0          0.0          0.0   
160 -0.709045     0.0  ...     0.0          1.0          0.0          0.0   
147 -0.709045     0.0  ...     0.0          1.0          0.0          0.0   
19  -0.709045     0.0  ...     0.0          1.0          0.0          0.0   

     slope_1.0  slope_2.0  slope_3.0  thal_3.0  thal_6.0  thal_7.0  
66         0.0        1.0        0.0       1.0       0.0       0.0  
105        1.0        0.0        0.0       0.0       0.0       1.0  
160        1.0        0.0        0.0       0.0       0.0       1.0  
147        1.0        0.0        0.0       1.0       0.0       0.0  
19         1.0        0.0        0.0       1.0       0.0       0.0  

[5 rows x 22 columns]

In [ ]:
model = Sequential()
model.add(Dense(units=32, activation='relu', input_dim=22))
model.add(Dense(units=64, activation='relu'))
model.add(Dense(units=1, activation='sigmoid'))
model.compile(loss='binary_crossentropy', optimizer='sgd',metrics=['accuracy'])



c:\Users\Sabiha\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:88: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [192]:
history = model.fit(
    X_train,
    y_train,
    batch_size=32,
    epochs=100,
    # We pass some validation for
    # monitoring validation loss and metrics
    # at the end of each epoch
    validation_data=(X_test, y_test),
)

Epoch 1/100
7/7 ━━━━━━━━━━━━━━━━━━━━ 2s 81ms/step - accuracy: 0.5597 - loss: 0.6654 - val_accuracy: 0.6533 - val_loss: 0.6233
Epoch 2/100
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 16ms/step - accuracy: 0.6129 - loss: 0.6540 - val_accuracy: 0.7333 - val_loss: 0.6126
Epoch 3/100
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step - accuracy: 0.6681 - loss: 0.6309 - val_accuracy: 0.7467 - val_loss: 0.6024
Epoch 4/100
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 21ms/step - accuracy: 0.7038 - loss: 0.6132 - val_accuracy: 0.7600 - val_loss: 0.5926
Epoch 5/100
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step - accuracy: 0.6950 - loss: 0.6068 - val_accuracy: 0.8000 - val_loss: 0.5832
Epoch 6/100
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 16ms/step - accuracy: 0.7230 - loss: 0.6023 - val_accuracy: 0.8000 - val_loss: 0.5740
Epoch 7/100
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 17ms/step - accuracy: 0.7434 - loss: 0.5977 - val_accuracy: 0.8000 - val_loss: 0.5651
Epoch 8/100
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step - accuracy: 0.7342 - loss: 0.5902 - val_accuracy: 0.8133 - val_loss:

In [193]:
test_loss, test_acc = model.evaluate(X_test, y_test)
print(f"Test accuracy: {test_acc:.4f}")

3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.8630 - loss: 0.3079
Test accuracy: 0.8667
